### ISSUE TITLE
Windows compilation issue - unable to find correct definitions of _InterlockedExchangeAdd

### ISSUE THREAD

**Issue by gsharma-ad** [openssl/openssl#21080]
OpenSSL 3.0.8

Platform: **Windows**
Compiler: **MCVC 2012** (I know it is very old)

I am getting below error while compiling an application with OpenSSL. 
```
connection.cc
C:\Shashank\cadp-c\src\build\Windows_NT_6.3.9600_64\include\internal/refcount.h(137) : error C2665: '_InterlockedExchangeAdd' : none of the 4 overloads could convert all the argument types
        c:\Program Files (x86)\Microsoft SDKs\Windows\v7.0A\include\winnt.h(2525): could be 'LONG _InterlockedExchangeAdd(volatile LONG *,LONG)'
        c:\Program Files (x86)\Microsoft SDKs\Windows\v7.0A\include\winbase.h(15323): or       'unsigned int _InterlockedExchangeAdd(volatile unsigned int *,unsigned int)'
        c:\Program Files (x86)\Microsoft SDKs\Windows\v7.0A\include\winbase.h(15343): or       'unsigned long _InterlockedExchangeAdd(volatile unsigned long *,unsigned long)'
        c:\Program Files (x86)\Microsoft SDKs\Windows\v7.0A\include\winbase.h(15365): or       'unsigned __int64 _InterlockedExchangeAdd(volatile unsigned __int64 *,unsigned __int64)'
        while trying to match the argument list '(volatile int *, int)'
C:\Shashank\cadp-c\src\build\Windows_NT_6.3.9600_64\include\internal/refcount.h(144) : error C2665: '_InterlockedExchangeAdd' : none of the 4 overloads could convert all the argument types
        c:\Program Files (x86)\Microsoft SDKs\Windows\v7.0A\include\winnt.h(2525): could be 'LONG _InterlockedExchangeAdd(volatile LONG *,LONG)'
        c:\Program Files (x86)\Microsoft SDKs\Windows\v7.0A\include\winbase.h(15323): or       'unsigned int _InterlockedExchangeAdd(volatile unsigned int *,unsigned int)'
        c:\Program Files (x86)\Microsoft SDKs\Windows\v7.0A\include\winbase.h(15343): or       'unsigned long _InterlockedExchangeAdd(volatile unsigned long *,unsigned long)'
        c:\Program Files (x86)\Microsoft SDKs\Windows\v7.0A\include\winbase.h(15365): or       'unsigned __int64 _InterlockedExchangeAdd(volatile unsigned __int64 *,unsigned __int64)'
        while trying to match the argument list '(volatile int *, int)'
```
Previously this issue was getting solved by modifying refcount.h as below -:
```
from _InterlockedExchangeAdd(val, -1); to _InterlockedExchangeAdd((long volatile*) val, -1);
```
Now I want to overcome this issue without modifying refcount.h.
Please suggest as OpenSSL is pretty new thing for me.

**Comment by gsharma-ad** [comment]
Configure cmd:
          **perl Configure VC-WIN64A no-asm no-idea  no-rc5  no-cast ;**
Then nmake.

We are using cygwin on windows to build our application.

**Comment by gvanem** [comment]
> We are using cygwin on windows to build our application.

Cygwin does not produce paths with `\\` slashes. 
A Cygwin `perl Configure VC-WIN64A` would cause an error if you do that. How did you **not** get that error?

But granted, the Perl (a language for old men) issues in OpenSSL are staggering. 
E.g. right now (from `master`). I have to use:
 * a Strawberry Perl for `configdata.pm`.
 * But a Cygwin Perl for the rest due to the new crap in `util/perl/OpenSSL/paramnames.pm` .
 
Go figure!

**Comment by gsharma-ad** [comment]
We are using cygwin for make command only for our application and perl installed on Windows machine outside Cygwin.
Compiler is MSVC 2012.

**Comment by t8m** [comment]
> * a Strawberry Perl for `configdata.pm`.
> 
> * But a Cygwin Perl for the rest due to the new crap in `util/perl/OpenSSL/paramnames.pm` .

We do not use Cygwin Perl in Windows CI runs and the build passes. How could it happen that you need it?

**Comment by gvanem** [comment]
>  How could it happen that you need it?

Because of this [issue](https://github.com/openssl/openssl/issues/21152) I just created.

**Comment by gsharma-ad** [comment]
Any update how fix issue mentioned in the description?
We are using cygwin for make command only and using MSVC 2010 to build our application.

Configure cmd:
**perl Configure VC-WIN64A no-asm no-idea no-rc5 no-cast** ;
Then nmake.

**Comment by gsharma-ad** [comment]
Fixed in 3.1 and 3.0 branches.

**Comment by keithc-ca** [comment]
The solution seems questionable to me; see https://github.com/openssl/openssl/pull/23087/files#r1447654261.

**Comment by gsharma-ad** [comment]
I am facing this problem with OpenSSL 3.5 also. PR  https://github.com/openssl/openssl/pull/23087 shall be merged into 3.5 also.

**Comment by tanjahummel** [comment]
@gsharma-ad I'm also facing this problem with OpenSSL 3.5. Are there any plans to merge the fix into the 3.5 branch soon?
